### ISSUE TITLE
3D - Stereoscopic 3D Support

### ISSUE THREAD

**Issue by mkst** [mkst/sm64-port#17]
# Introduction
This issue has been created to track the progress of adding stereoscopic 3D to the 3ds-port.

![image](https://user-images.githubusercontent.com/22226349/90907995-b1424d00-e3cb-11ea-8591-fc7df8e84f20.png)

# Progress:

3D is applied via applying `Mtx_PerspStereoTilt` to the projection applied in the vertex shader. Not everything should have this effect, otherwise 2d things with high (low?) Z-index pop out and poke your eyes.

The current approach to determining 2D is based on an extra call to 'set 2d mode' at specific points in the SM64 code. We also check that we are drawing something with **2** triangles; both of these conditions have to be met for the item _not_ to be drawn in 3D.

## 3D
Apply 3D effect to everything in the 3D space.

 - [x] Rudimentary 3D using `Mtx_PerspStereoTilt` from [citro3d](https://github.com/devkitPro/citro3d/blob/52cdcb351a6697d34cc4004cf7ebc7392cd19710/source/maths/mtx_perspstereotilt.c)
 - [x] Snow effect (potentially others)
 - [ ] Adjust focal length based on distance between mario+camera
 - [x] Skybox

## 2D
We need to work out what is 2D and ensure we are not applying a 3D effect to that.

 - [x] HUD (`src/game/hud.c`)
 - [x] PAUSE screen ('src/game/print.c')
 - [x] Signposts  ('src/game/print.c')
 - [x] Star select screen (`src/menu/star_select.c`)
 - [x] Peach letter (`src/game/ingame_menu.c`)
 - [x] Cannon overlay
 - [x] Level/Area transitions
 - [ ] File select screen; it would be nice if whole screen was treated as 2D
 - [x] `TM` and `(C) Nintendo` on first intro screen
 - [ ] Goddard background (somewhat fixed on 3ds-minimap branch)
 - [x] Cutscene captions during ending are missing ("mario!")
 - [ ] more items TBD

## Extra

 - [x] Determine why an FOV is `54.0f` rather than `45.0f` is required to maintain ~~same~~ similar perspective to 2D mode; can we apply some viewport shifting?
 - [x] Determine why extra args need to be set in perspective matrix in order to render anything
 - [x] Apply scissoring to viewport for left/right eye after perspective applied (depends how strong 3D effect is)

# Notes:
 - Setting POV to 90 shows that the STEREO button and Goddard moustache are being rendered with higher Z level than everything else, so presumably are being clipped... why?

# Addendum:

The closed-source port by borb is simply doing:
```c
    glMatrixMode(GL_PROJECTION);
    glLoadMatrixf(&(*gsp_mtxf_projection)[0][0]);
    glTranslatef(gsp_depth, 0.0F, 0.0F);
```

**Comment by ryotoko** [comment]
~~Draw distance is adversely affected and triangles on the periphery are being clipped
Skybox in certain areas seems to be 2D (first Bowser arena?)~~
Fixed

**Comment by mkst** [comment]
> Draw distance is adversely affected and triangles on the periphery are being clipped

Yup, this is tracked under "Apply scissoring..."

> Skybox in certain areas seems to be 2D (first Bowser arena?)

Interesting. I'll take a closer look - there's a lot I still need to understand.

**Comment by ryotoko** [comment]
Looking carefully it seems like all of the skyboxes are being rendered on the same depth level (0?) as the HUD is

**Comment by ryotoko** [comment]
![Untitled](https://user-images.githubusercontent.com/68757712/92665308-11395e80-f2bb-11ea-9d4f-e149d5278821.png)
Made a crude illustration here to be clearer about the skybox problem. The star represents the skybox texture, with the heart in the foreground and a tree somewhere in between. I don't know if you'd notice this if you're using red/blue 3D glasses, but on the actual system it definitely ruins the effect. The Goddard background has the same problem: it appears less "deep" than the face. 

Other than that and the scissoring, the current build is remarkably serviceable.

**Comment by mkst** [comment]
@ryotoko I don't really know what to do with the skybox, it's just a 2D background...

**Comment by ryotoko** [comment]
Can you control how deep it is? That's the only thing that needs to be fixed. Its current Z value has to be coming from somewhere. Easier said than done I guess

**Comment by mkst** [comment]
> the skybox in sm64 is an ortho projected 2d plane
[source](https://discord.com/channels/601153657966428193/615010554809155604/755153023298175067).

**Comment by ryotoko** [comment]
I'm not disputing that it's a 2D Plane - it is. The problem is how deep in this 3D space that plane is occurring. Right now it's almost at the same level as the HUD elements (slightly behind), so you see the skybox popping out at you in the same way that you see the HUD elements popping out, whereas the game environments and actors occur much deeper inside that space, deeper than the skybox. "deeper than the skybox" is not something that should exist in a 3D space, because everything that's behind the skybox should be invisible. 
I've done my best to describe this problem so I'm just going to leave it at this.

**Comment by mkst** [comment]
Haha - I'm completely new to this. Have you tried the `3ds-minimap` branch?
I have moved the hud to the bottom screen. I basically lack the ability to
"fix" the skybox. I don't even understand what needs to be done.

**Comment by ryotoko** [comment]
Well you've done a great job implementing 3D. I certainly know far less about this.

I noticed this in skybox.c:
 * The skybox is always drawn behind everything, because in the level's geo script, the skybox is drawn
 * first, in a display list with the Z buffer disabled
And it makes sense because you want to see the full skybox at all times. If the skybox was further into the Z dimension, then it would effectively get smaller and you'd eventually see where the texture edges loop together, which is not something you should see especially on the top and bottom

So the solution isn't to make the plane deeper, that would just result in a distant, tiled skybox texture which is still wrong. What you really need to do is offset the X-axis scaling when drawing the skybox for each eye's display, with the left eye being a bit more to the left and the right eye being a bit more to the right. I have no idea what this value would be and ideally it would change with intensity in response to the 3D slider. But it would be a constant, that just simulates the depth of the skybox plane without actually making it deeper.

**Comment by ryotoko** [comment]
As it stands, the current "depth" of the skybox, slightly deeper than the HUD, is coming from the slightly different camera positions from the left and right eye displays, because the camera position is what's determining what portion of the skybox texture is being drawn. So you'd want to add 3D-mode conditionals that subtract some amount from the left eye X value and add some amount to the right eye X value. I mean, that's the logic of it at least, I'm pretty confident in that. I just don't have the slightest clue how to actually do it

**Comment by oneup03** [comment]
I haven't looked at the code, but maybe some of the techniques that the Nvidia 3DVision community uses to stereoize PC games would prove helpful to fixing the skybox. Here is a wiki guide that talks about some of the basic shader hacking: [http://wiki.bo3b.net/index.php?title=Bo3b%27s_School_for_Shaderhackers](url) The prime directive section talks about this formula: clipPos.x += EyeSign * Separation * ( clipPos.w – Convergence )  I'm not sure if this could be applied to the skybox easily. Some of the developers with more knowledge might be able to help on their Discord: [https://discord.gg/tXHYhNK](url)

**Comment by mkst** [comment]
The stereoscopic effect is applied in the [gfx_citro3d_draw_triangles_helper](https://github.com/mkst/sm64-port/blob/3ds-port/src/pc/gfx/gfx_citro3d.c#L729) function via calling:

```c
Mtx_PerspStereoTilt(&projection, fov, 1.0f , 0.1f, 10.0f, -iod, focalLen, false); // left eye
...
Mtx_PerspStereoTilt(&projection, fov, 1.0f, 0.1f, 10.0f, iod, focalLen, false); // right eye
```

This is applied to the scene immediately before it's rendered to screen.

I assume that as the skybox has no z-index it does not get the perspective applied, which is why it just appears 2D?

Out of interest, have either of you been able to get this working in citra? It crashes out before Goddard for me which makes testing laborious as I have to upload to my 3DS every build...

**Comment by ryotoko** [comment]
The skybox is currently being drawn in a slightly different X position for both eyes, so at the very least the skybox program has to be running twice already, it's just not offsetting the X position enough for the other eye to achieve the illusion of depth. The real problem is making the skybox program aware of which eye is being calculated. If this were all one long main.c I think I could fix it, but my knowledge is so rudimentary I don't know how to reference the gfx functions from within the skybox program. 

But it would look something like this in [https://github.com/mkst/sm64-port/blob/c64de52b7c54a54235a6d26fc9c3a3d6e1190e15/src/game/skybox.c#L153](url):
```
calculate_skybox_scaled_x(s8 player, f32 fov) {
    f32 yaw = sSkyBoxInfo[player].yaw;
    #ifdef ENABLE_N3DS_3D_MODE
    if ((gGfx3DSMode == GFX_3DS_MODE_NORMAL || gGfx3DSMode == GFX_3DS_MODE_AA_22) && gSliderLevel > 0.0f) 
    // and if "C3D_FrameDrawOn(gTargetRight)" 
    {
    f32 xCorrect = (yaw / 65536.0) - 0.1;
    if (xCorrect < 0) { xCorrect = 1 + xCorrect; }
    f32 yawScaled = SCREEN_WIDTH * (360 / fov) * xCorrect;
    }
    else
    f32 yawScaled = SCREEN_WIDTH * 360.0 * yaw / (fov * 65536.0);
```

Basically, the effect would be kicking the yaw of the camera for the right eye 36 degrees counter-clockwise, shifting the skybox for the right eye some amount to the right. 36 degrees (0.1 of a 0~1 circle) is probably too much, but it's a starting point and this value could later be adjusted with respect to gSliderLevel

Sorry this is the limit of my usefulness, if at all

**Comment by mkst** [comment]
So, the draw_triangles_helper is what sets up the left/right eye when rendering the scene; it won't help to make those changes to the skybox.

It sounds like the skybox needs to be considered 3D (i.e. via adding a z-index etc), once that's done, the stereo shift will sort out the rest.

In terms of "what I did before", there was a bug where I wasn't flushing the scene before setting 2d/not 2d.

**Comment by ryotoko** [comment]
That's what I got to thinking too. I made skyboxes 3D again, and made this change to [line 237 of skybox.c](https://github.com/mkst/sm64-port/blob/7a2e471deaa11259b5038e2b579ab657f8948822/src/game/skybox.c#L237):

```c
        make_vertex(verts, 0, x, y, -3, 0, 0, sSkyboxColors[colorIndex][0], sSkyboxColors[colorIndex][1],
                    sSkyboxColors[colorIndex][2], 255);
        make_vertex(verts, 1, x, y - SKYBOX_TILE_HEIGHT, -3, 0, 31 << 5, sSkyboxColors[colorIndex][0], sSkyboxColors[colorIndex][1],
                    sSkyboxColors[colorIndex][2], 255);
        make_vertex(verts, 2, x + SKYBOX_TILE_WIDTH, y - SKYBOX_TILE_HEIGHT, -3, 31 << 5, 31 << 5, sSkyboxColors[colorIndex][0],
                    sSkyboxColors[colorIndex][1], sSkyboxColors[colorIndex][2], 255);
        make_vertex(verts, 3, x + SKYBOX_TILE_WIDTH, y, -3, 31 << 5, 0, sSkyboxColors[colorIndex][0], sSkyboxColors[colorIndex][1],
                    sSkyboxColors[colorIndex][2], 255);
```
Z value changed from -1 to -3. This effectively fixes the depth of the skybox problem, and has no effect on its scaling, so it's harmless in 2D mode.
Negative values draw it further away and it disappears completely after -3, presumably because of draw distance/render range culling. There's also minor clipping on the edges, either because of the tile drawing logic needs revised, or also because of the render range, similar to scissoring problem. Not perfect but it's way better now than 2D skybox

**Comment by ryotoko** [comment]
Sadly this seems to introduce some frame stuttering in 3D mode, I think the skybox tile logic is the culprit.

**Comment by 01y** [comment]
Another improvement would be to increase the "FOV" during frustum culling. Currently, one can see blank regions at the far left and right sides of the screen in many scenes, with polygons popping in and out visibly, due to the expanded viewing range brought on by the 3D mode.

**Comment by ryotoko** [comment]
Extremely minor but, only in 3D mode: during the ending cutscene, when the star turns the whole screen white briefly before Peach appears from the window, this turns the whole screen white, including the letterbox, which is supposed to remain black. The ending is pretty much the only time the game uses viewports and scissoring aside from the default 320x240, but I think the problem is there somewhere. But why 3D mode only?

**Comment by ryotoko** [comment]
~~3D mode bug: Toggling hi score/my score in the score menu in file select crashes the game~~
Fixed..
